### PR TITLE
Add prospect model and API functionality

### PR DESCRIPTION
**Pull request by airensurzyn**
Ref : https://docs.google.com/document/d/158R8wHdpWKpUmWExdoR5cKSQa8_UANnDiibPDyDQaug/edit

for details

### CODE REVIEW

File: server/models/Prospect.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * 
+ * We will want to make the 'owned_by' field a required field once we have workable user_id's
+ * 
+ **/
+
+const ProspectSchema = new Schema({
+    first_name: { type: String, required: true},
+    last_name: { type: String, required: true},
+    email: { type: String, required: true, index: true},
+    owned_by: {type: mongoose.Schema.Types.ObjectId, ref: 'User', required : false},
```
Comment (skassam21) [inline server/models/Prospect.js]: I think I wanted to index owned_by as opposed to email actually - we are going to be doing a lot of fetching all prospects by owned_by as opposed to email

Also, i mentioned this in Darren's PR too, we want to use camelCase for the fields names - see https://intellipaat.com/community/28608/what-are-naming-conventions-for-mongodb

File: server/validation/prospect.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  let errors = {};
+
+// Convert empty fields to an empty string so we can use validator functions
+  data.first_name = !isEmpty(data.first_name) ? data.first_name : "";
+  data.last_name = !isEmpty(data.last_name) ? data.last_name : "";
+  data.email = !isEmpty(data.email) ? data.email : "";
+  data.status = !isEmpty(data.status) ? data.status : "";
+
+// Name checks
+  if (Validator.isEmpty(data.first_name)) {
```
Comment (skassam21) [inline server/validation/prospect.js]: Do we need these checks if above you are setting them to "" if they are not empty?

File: server/routes/api/prospects.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  if (!isValid) {
+    return res.status(400).json(errors);
+}
+
+  Prospect.findById(id, function(err, prospect) {
+    if (!prospect) {
+      res.status(404).send({id : `Prospect with id ${id} is not found`});  
+    } else {
+      //if updating email field, check if Prospect with new email alredy exists
+      if(req.body.email!==prospect.email) {
```
Comment (skassam21) [inline server/routes/api/prospects.js]: spacing around !==, make sure we have prettier.io installed across all of you to have consistent styling

File: server/models/Prospect.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+const mongoose = require("mongoose");
+const Schema = mongoose.Schema;
+
+
+var Status = Object.freeze({
```
Comment (skassam21) [inline server/models/Prospect.js]: avoid var, use const instead

Comment (skassam21) [reply server/models/Prospect.js]: I feel like as a constant it should be STATUS instead

**Review by skassam21 — COMMENTED** [review]
The routes generally look good - let's talk about next steps in building the route for CSV uploads - we'd probably want to design that later when you start playing around with uploading on the front-end

**Comment by airensurzyn** [comment]
Alright, thanks Shums. I'll make these changes and re-push

**Comment by airensurzyn** [comment]
@skassam21 PR fixes are in

File: server/app.js [hunk 5]
```diff
@@ -29,6 +34,7 @@ app.use(express.static(join(__dirname, "public")));
 
 app.use("/", indexRouter);
 app.use("/ping", pingRouter);
+app.use("/api/prospects", prospectRouter);
```
Comment (Disquietude) [inline server/app.js]: As discussed in the stand-up, we should import our routers into `routes/index.js` and have that router `.use()` them

**Comment by airensurzyn** [comment]
Merged the new dev changes. I believe this is ready to merge to dev as well if no one has any further comments. Will wait until morning to do so.

File: server/routes/index.js [hunk 6]
```diff
@@ -1,8 +1,12 @@
 const express = require("express");
 const router = express.Router();
+const prospectRouter = require("./api/prospects");
```
Comment (Disquietude) [inline server/routes/index.js]: Minor nitpick: `const prospects` instead of `const prospectRouter`

File: package-lock.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+  "lockfileVersion": 1
+}
```
Comment (Disquietude) [inline package-lock.json]: This file shouldn't exist

File: server/validation/prospect.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+const Validator = require("validator");
+
+const { STATUS } = require("../models/Prospect")
+
+module.exports = function validateProspectInput(data) {
+  const errors = {};
+
+// Name checks
+  if ((!data.firstName)) {
```
Comment (Disquietude) [inline server/validation/prospect.js]: Unnecessary double brackets

File: server/validation/prospect.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+// Status checks  
+if (!data.status) {
+    errors.status = "Status is required";
+} else if (!Object.values(STATUS).includes(data.status)) {
+    errors.status = "Invalid Prospect status";
+}
+
+return {
+    errors,
+    isValid: (!Object.keys(errors).length > 0)
```
Comment (Disquietude) [inline server/validation/prospect.js]: No need for brackets here

**Review by Disquietude — APPROVED** [review]
Looks good overall, just needs some minor fixes

**Pull request merged**